uptime/checker.go

```diff
@@ -0,0 +1,416 @@
+package uptime
+
+import (
+	"context"
+	"sync"
+	"fmt"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/filecoin-project/lotus/api/v0api"
+	"github.com/filecoin-project/go-address"
+	chainTypes "github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+	"github.com/libp2p/go-libp2p-core/host"
+	libp2pMultiaddr "github.com/multiformats/go-multiaddr"
+	peerstore "github.com/libp2p/go-libp2p-core/peer"
+)
+
+var log = logging.Logger("uptime")
+
+const NEW_CHECKER = 2
+const REPORT_CHECKER = 8
+
+// UptimeChecker maintains the uptime of member nodes
+type UptimeChecker struct {
+	api v0api.FullNode
+
+	self ActorID
+	actorAddress address.Address
+	
+	checkerAddresses []MultiAddr
+	nodeAddresses map[ActorID]map[MultiAddr]HealtcheckInfo
+
+	// libp2p ping related
+	node host.Host
+	ping *ping.PingService
+
+	rwLock sync.RWMutex
+	stop bool
+}
+
+func NewUptimeChecker(
+	api v0api.FullNode,
+	actorAddress string,
+	checkerAddresses []MultiAddr,
+	self ActorID,
+	node host.Host,
+	ping *ping.PingService,
+) (UptimeChecker, error) {
+	addr, err := address.NewFromString(actorAddress)
+	if err != nil {
+		return UptimeChecker{}, nil
+	}
+	return UptimeChecker {
+		api: api,
+
+		self: self,
+		actorAddress: addr,
+
+		checkerAddresses: checkerAddresses,
+		nodeAddresses: make(map[ActorID]map[MultiAddr]HealtcheckInfo),
+
+		node: node,
+		ping: ping,
+
+		stop: false,
+	}, nil
+}
+
+func (u *UptimeChecker) Start(ctx context.Context) error {
+	hasRegistered, err := u.HasRegistered(ctx)
+	if err != nil {
+		return err
+	}
+
+	if !hasRegistered {
+		if err := u.Register(ctx); err != nil {
+			return err
+		}
+	} else {
+		log.Infow("already registered with the actor, skip register")
+	}
+
+	go func() {
+		u.processReportedCheckers(ctx)
+	}()
+
+	// go func() {
+	// 	u.monitorMemberNodes(ctx)
+	// }()
+
+	go func() {
+		u.monitorCheckerNodes(ctx)
+	}()
+
+	return nil
+}
+
+// HasRegistered checks if the current checker has already registered itself in the actor
+func (u *UptimeChecker) HasRegistered(ctx context.Context) (bool, error) {
+	s, err := Load(ctx, u.api, u.actorAddress, u.self)
+	if err != nil {
+		return false, err
+	}
+	return s.HasRegistered(u.self)
+}
+
+// Register registers the current checker to actor
+func (u *UptimeChecker) Register(ctx context.Context) error {
+	log.Infow("has yet to be registered with the actor, register now")
+
+	peerID, err := u.api.ID(ctx);
+	if err != nil {
+		return nil
+	}
+	
+	log.Infow("register with peer id", "peerID", peerID.String())
+
+	params, err := encodeJson(NodeInfo {
```

Comment:
Have you tried this? Does sending the serialized params as a JSON directly works? I would expect that they need to be CBOR serialized.

Reply:
yeah, I tried, it works. Because on the actor, it is decoding json as well. As long as both sides are aligned, should be ok. Reason I used json is because it's easier to test on cli. I can change to CBOR as well